File: java/AtomTrivalent.java
/*
 *  Copyright (c) 2014-2015 Bruce Twite
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Trivalent atom, either true, false or maybe.
 */
package yaga.core;

/**
 *
 * @author Bruce
 */
public abstract class AtomTrivalent extends AtomicList
{
	static public AtomTrivalent TRUE =		new True();
	static public AtomTrivalent FALSE =		new False();
	static public AtomTrivalent UNKNOWN =	new Unknown();
	
	private AtomTrivalent(Symbol.Reserved trit)
		{ _trit = trit; }
	
	private final Symbol.Reserved _trit;
	
	public int getValue()
		{ return (_trit.ordinal()); }
	public Symbol.Reserved getTrit()
		{ return (_trit); }
	
	@Override
	public AtomTrivalent zeroValue()
		{ return (FALSE); }
	
	@Override
	public final boolean isTrivalent()
		{ return (true); }

	@Override
	public void print(StringBuilder sb) 
		{ Symbol.printReserved(sb, _trit); }

	@Override
	public void xprint(StringBuilder sb) 
		{ Symbol.xprintReserved(sb, _trit, "Trivalent"); }

	@Override
	public List add(Context ctxt, List e)
		{ return (UNKNOWN); }
	@Override
	public List sub(Context ctxt, List e)
		{ return (UNKNOWN); }
	@Override
	public List mul(Context ctxt, List e)
		{ return (UNKNOWN); }
	@Override
	public List div(Context ctxt, List e)
		{ return (UNKNOWN); }
	@Override
	public List rem(Context ctxt, List e)
		{ return (UNKNOWN); }
	
	static private class True extends AtomTrivalent
	{
		private True()
			{ super(Symbol.Reserved.TRUE); }
		
		@Override
		public final boolean isTrue()
			{ return (true); }

		@Override
		public final List trueSelect(List eTrue, List eElse)
			{ return (eTrue); }
		@Override
		public final List falseSelect(List eFalse, List eElse)
			{ return (eElse); }
		@Override
		public final List unknownSelect(List eUnknown, List eElse)
			{ return (eElse); }
		@Override
		public List trueFalseSelect(List eTrue, List eFalse, List eElse)
			{ return (eTrue); }
		@Override
		public final List select(List eTrue, List eFalse, List eUnknown)
			{ return (eTrue); }
		@Override
		public List select(List eTrue, List eFalse, List eUnknown, List eElse)
			{ return (eTrue); }
	
		@Override
		public AtomTrivalent neg(Context ctxt)
			{ return (FALSE); }
	}
	
	static private class False extends AtomTrivalent
	{
		private False()
			{ super(Symbol.Reserved.FALSE); }

		@Override
		public final boolean isFalse()
			{ return (true); }

		@Override
		public final List trueSelect(List eTrue, List eElse)
			{ return (eElse); }
		@Override
		public final List falseSelect(List eFalse, List eElse)
			{ return (eFalse); }
		@Override
		public final List unknownSelect(List eUnknown, List eElse)
			{ return (eElse); }
		@Override
		public List trueFalseSelect(List eTrue, List eFalse, List eElse)
			{ return (eFalse); }
		@Override
		public final List select(List eTrue, List eFalse, List eUnknown)
			{ return (eFalse); }
		@Override
		public List select(List eTrue, List eFalse, List eUnknown, List eElse)
			{ return (eFalse); }
	
		@Override
		public AtomTrivalent neg(Context ctxt)
			{ return (TRUE); }
	}
	
	static private class Unknown extends AtomTrivalent
	{
		private Unknown()
			{ super(Symbol.Reserved.UNKNOWN); }
		
		@Override
		public final boolean isUnknown()
			{ return (true); }

		@Override
		public final List trueSelect(List eTrue, List eElse)
			{ return (eElse); }
		@Override
		public final List falseSelect(List eFalse, List eElse)
			{ return (eElse); }
		@Override
		public final List unknownSelect(List eUnknown, List eElse)
			{ return (eUnknown); }
		@Override
		public List trueFalseSelect(List eTrue, List eFalse, List eElse)
			{ return (eElse); }
		@Override
		public final List select(List eTrue, List eFalse, List eUnknown)
			{ return (eUnknown); }
		@Override
		public List select(List eTrue, List eFalse, List eUnknown, List eElse)
			{ return (eUnknown); }
	
		@Override
		public AtomTrivalent neg(Context ctxt)
			{ return (this); }
	}
}

File: AtomTrivalent.js
/**
 *  AtomTrivalent : @file
 *
 *  Trivalent atom, either true, false or maybe.
 */

var yc;
module.exports = {
	prototype: undefined,
	TRUE: undefined,
	FALSE: undefined,
	UNKNOWN: undefined,
	Initialise: _init,
};

function _init(y) {
	if (yc) return;
	yc = y;
	let _prot = Object.assign(Object.create(yc.List.prototype), {
		typeName: 'AtomTrivalent',
		setTrit(trit) {
			this._trit = trit;
		},
		getValue() {
			return (this._trit.ordinal);
		},
		getTrit() {
			return (this._trit);
		},
		zeroValue: () => _false,
		isTrivalent: () => true,
		print(sb) {
			yc.Symbol.printReserved(sb, this._trit);
		},
		xprint(sb) {
			yc.Symbol.xprintReserved(sb, this._trit, 'Trivalent');
		},
		add: (ctxt, e) => _unknown,
		sub: (ctxt, e) => _unknown,
		mul: (ctxt, e) => _unknown,
		div: (ctxt, e) => _unknown,
		rem: (ctxt, e) => _unknown,
	});

	let _true = Object.assign(Object.create(_prot), {
		typeName: 'TRUE',
		isTrue: () => true,
		trueSelect(eTrue, eElse) {
			return (eTrue);
		},
		falseSelect(eFalse, eElse) {
			return (eElse);
		},
		unknownSelect(eUnknown, eElse) {
			return (eElse);
		},
		trueFalseSelect(eTrue, eFalse, eElse) {
			return (eTrue);
		},
		selectTrueFalseUnknown(eTrue, eFalse, eUnknown) {
			return (eTrue);
		},
		selectTrueFalseUnknownElse(eTrue, eFalse, eUnknown, eElse) {
			return (eTrue);
		},
		neg(ctxt) {
			return (_false);
		}
	});
	_true.setTrit(yc.Symbol.Reserved.TRUE);

	let _false = Object.assign(Object.create(_prot), {
		typeName: 'FALSE',
		isFalse: () => true,
		trueSelect(eTrue, eElse) {
			return (eElse);
		},
		falseSelect(eFalse, eElse) {
			return (eFalse);
		},
		unknownSelect(eUnknown, eElse) {
			return (eElse);
		},
		trueFalseSelect(eTrue, eFalse, eElse) {
			return (eFalse);
		},
		selectTrueFalseUnknown(eTrue, eFalse, eUnknown) {
			return (eFalse);
		},
		selectTrueFalseUnknownElse(eTrue, eFalse, eUnknown, eElse) {
			return (eFalse);
		},
		neg(ctxt) {
			return (_true);
		}
	});
	_false.setTrit(yc.Symbol.Reserved.FALSE);

	let _unknown = Object.assign(Object.create(_prot), {
		typeName: 'UNKNOWN',
		isUnknown: () => true,
		trueSelect(eTrue, eElse) {
			return (eElse);
		},
		falseSelect(eFalse, eElse) {
			return (eElse);
		},
		unknownSelect(eUnknown, eElse) {
			return (eUnknown);
		},
		trueFalseSelect(eTrue, eFalse, eElse) {
			return (eElse);
		},
		selectTrueFalseUnknown(eTrue, eFalse, eUnknown) {
			return (eUnknown);
		},
		selectTrueFalseUnknownElse(eTrue, eFalse, eUnknown, eElse) {
			return (eUnknown);
		},
		neg(ctxt) {
			return (this);
		}
	});
	_false.setTrit(yc.Symbol.Reserved.UNKNOWN);

	module.exports.prototype = _prot;
	module.exports.TRUE = _true;
	module.exports.FALSE = _false;
	module.exports.UNKNOWN = _unknown;

	Object.freeze(module.exports);
}